Browserify should not create empty output files on error

When an error occurs, Browserify 5.11.2 creates an empty output file.  This should not happen, as it messes up `make`-based workflows.

```
$ cat >foo.js <<EOF
var quux = require('quux');
EOF

$ browserify foo.js -o bar.js
Error: Cannot find module 'quux' from '/private/tmp/foo'
    at /usr/local/lib/node_modules/browserify/node_modules/resolve/lib/async.js:50:17
    at process (/usr/local/lib/node_modules/browserify/node_modules/resolve/lib/async.js:119:43)
    at /usr/local/lib/node_modules/browserify/node_modules/resolve/lib/async.js:128:21
    at load (/usr/local/lib/node_modules/browserify/node_modules/resolve/lib/async.js:60:43)
    at /usr/local/lib/node_modules/browserify/node_modules/resolve/lib/async.js:66:22
    at /usr/local/lib/node_modules/browserify/node_modules/resolve/lib/async.js:21:47
    at Object.oncomplete (fs.js:107:15)

$ ls -l | grep bar.js
-rw-r--r--  1 mietek  wheel   0 13 Sep 14:53 bar.js
```

--- a/bin/cmd.js
+++ b/bin/cmd.js
@@ -54,16 +54,23 @@ if (b.argv.list) {
 
 var bundle = b.bundle();
 bundle.on('error', errorExit);
+bundle.on('end', successExit);
 
+var tmpfile;
 var outfile = b.argv.o || b.argv.outfile;
 if (outfile) {
-    bundle.pipe(fs.createWriteStream(outfile));
+    // we'll output to a temp file within same filesystem, then atomically overwrite outfile once successful
+    tmpfile = outfile + ".tmp-browserify-" + Math.random().toFixed(20).slice(2)
+    bundle.pipe(fs.createWriteStream(tmpfile));
 }
 else {
     bundle.pipe(process.stdout);
 }
 
 function errorExit(err) {
+    if (tmpfile) fs.unlink(tmpfile, function (err) {
+      if (err) /* no-op, we're already exiting unhappily… */;
+    });
     if (err.stack) {
         console.error(err.stack);
     }
@@ -71,4 +78,10 @@ function errorExit(err) {
         console.error(String(err));
     }
     process.exit(1);
+}
+
+function successExit() {
+  if (tmpfile && outfile) fs.rename(tmpfile, outfile, function (err) {
+    if (err) errorExit(err);
+  });
 }

